packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:mojo/core.dart' as core;
+
+import 'asset_bundle.dart';
+
+/// Base class for asset resolvers.
+abstract class AssetResolver {
+  /// Return a resolved asset key for the asset named [name].
+  Future<String> resolve(String name);
+}
+
+/// Builder for a particular AssetResolver.
+///
+/// Allows AssetResolvers to be specified in widget build() functions.
+/// Subclasses should override operator== and hashCode so that equivalent
+/// configurations appear to be equal. Builder objects should also be
+/// cheap to construct.
+abstract class AssetResolverBuilder<T extends AssetResolver> {
+  const AssetResolverBuilder();
+  T buildResolver();
+}
+
+/// Base class for resolvers that use the asset manifest to retrieve a list
+/// of asset variants to choose from.
+abstract class VariantAssetResolver extends AssetResolver {
+  VariantAssetResolver({ this.bundle });
+  final AssetBundle bundle;
+  Map<String, List<String>> _assetManifest = null;
+
+  Future<String> resolve(String name) async {
+    if (_assetManifest == null) {
+      String json = await bundle.loadString("AssetManifest.json");
+      _assetManifest = JSON.decode(json);
```

Comment:
We should have some error handling here.  What if `AssetManifest.json` doesn't exist in the bundle or doesn't parse in the expected way?  Probably we should catch the exception and log an error.